language/move-lang/src/cfgir/cfg.rs

```diff
@@ -70,31 +87,145 @@ impl<'a> BlockCFG<'a> {
+
+    fn commands<'s>(&'s self, label: Label) -> Box<dyn Iterator<Item = &'s Command> + 's> {
+        Box::new(self.block(label).iter())
+    }
+
+    fn commands_mut<'s>(
+        &'s mut self,
+        label: Label,
+    ) -> Box<dyn Iterator<Item = &'s mut Command> + 's> {
+        Box::new(self.block_mut(label).iter_mut())
+    }
+
+    fn num_blocks(&self) -> usize {
+        self.blocks.len()
+    }
+
+    fn start_block(&self) -> Label {
+        self.start
+    }
+}
+
+//**************************************************************************************************
+// BlockCFG
+//**************************************************************************************************
+
+#[derive(Debug)]
+pub struct ReverseBlockCFG<'a> {
+    start: Label,
+    blocks: &'a mut Blocks,
+    successor_map: BTreeMap<Label, BTreeSet<Label>>,
+    predecessor_map: BTreeMap<Label, BTreeSet<Label>>,
+}
+
+impl<'a> ReverseBlockCFG<'a> {
+    pub fn new(forward_cfg: &'a mut BlockCFG, infinite_loop_starts: BTreeSet<Label>) -> Self {
+        let blocks: &'a mut Blocks = &mut forward_cfg.blocks;
+        let forward_successors = &forward_cfg.successor_map;
+        let forward_predecessor = &forward_cfg.predecessor_map;
+        let end_blocks = {
+            let mut end_blocks = BTreeSet::new();
+            for (lbl, successors) in forward_successors {
+                let loop_start_successors = successors
+                    .iter()
+                    .filter(|l| infinite_loop_starts.contains(l));
+                for loop_start_successor in loop_start_successors {
+                    if lbl >= loop_start_successor {
+                        end_blocks.insert(*lbl);
+                    }
+                }
+            }
+            for (lbl, block) in blocks.iter() {
+                let last_cmd = block.back().unwrap();
+                if last_cmd.value.is_divergent() {
+                    end_blocks.insert(*lbl);
+                }
+            }
+            end_blocks
+        };
+
+        let mut successor_map = forward_predecessor.clone();
+        let mut predecessor_map = forward_successors.clone();
+
+        let terminal = Label(blocks.len());
+        assert!(!blocks.contains_key(&terminal));
+        for terminal_successor in &end_blocks {
+            predecessor_map
+                .entry(*terminal_successor)
+                .or_insert_with(BTreeSet::new)
+                .insert(terminal);
+        }
+        successor_map.insert(terminal, end_blocks);
+        blocks.insert(terminal, BasicBlock::new());
+
+        Self {
+            start: terminal,
+            blocks,
+            successor_map,
+            predecessor_map,
+        }
+    }
+
+    pub fn block(&self, label: Label) -> &BasicBlock {
+        self.blocks.get(&label).unwrap()
+    }
+
+    pub fn block_mut(&mut self, label: Label) -> &mut BasicBlock {
+        self.blocks.get_mut(&label).unwrap()
     }
+}
+
+impl<'a> Drop for ReverseBlockCFG<'a> {
```

Comment:
Not sure if is is the best way to do this, but it seemed to be the least brittle.

Reply:
I think taking `forward_cfg: &'a BlockCFG` above and cloning `forward_cfg.blocks` would be even less brittle. That is, unless you actually want to mutate the `blocks` of the forward CFG later, which I could see being useful for optimization.

This `Drop` trick is interesting as a way of building a temporary "view" in Rust. You could imagine mutating `successor_map` and `predecessor_map` above instead of cloning them, then undoing those mutations in `Drop` as well.

Reply:
> That is, unless you actually want to mutate the blocks of the forward CFG later, which I could see being useful for optimization.

Originally, I had planned on mutating, but I'm unsure what I'll do exactly there now.

> This Drop trick is interesting as a way of building a temporary "view" in Rust. You could imagine mutating successor_map and predecessor_map above instead of cloning them, then undoing those mutations in Drop as well.

I am afraid of over relying on drop too much, but maybe it will be fine since I will be taking a mutable reference (and thus taking ownership). It's something to ponder, but I might stick with using `Drop` as little as possible right now

Reply:
I actually ended up taking references to both, and then cleaning up with `Drop`. The implementation ended up looking pretty nice